src/addrman.h

```diff
@@ -527,8 +527,23 @@ friend class CAddrManTest;
             CAddrInfo &info = mapInfo[n];
             int bucket = entryToBucket[n];
             int nUBucketPos = info.GetBucketPosition(nKey, true, bucket);
-            if (format >= Format::V2_ASMAP && nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && vvNew[bucket][nUBucketPos] == -1 &&
-                info.nRefCount < ADDRMAN_NEW_BUCKETS_PER_ADDRESS && serialized_asmap_version == supplied_asmap_version) {
+
+            // We already checked that the file format is not "too new" and we can parse it (see lowest_compatible above).
+            // Here, check that the file is compatible by other means:
+            // - constants haven't been changed incompatibly
+            // - asmap has not changed
+            // - bucket/position assignment hasn't been changed incompatibly
+            // If it's incompatible, reassign buckets/positions.
+
+            // asmap has changed in two cases:
+            // - either it wasn't supported by the addr file, but now asmap is supplied;
+            // - or it was supported, but we have a different asmap now (the lack of asmap counts as well)
+            const bool asmap_is_same = serialized_asmap_version == supplied_asmap_version;
+
+            const bool constants_are_compatible = nUBuckets == ADDRMAN_NEW_BUCKET_COUNT && info.nRefCount < ADDRMAN_NEW_BUCKETS_PER_ADDRESS;
+            const bool assignment_is_compatible = vvNew[bucket][nUBucketPos] == -1;
```

Comment:
I still think these assignments should happen outside the loop, since they're constant and don't depend on the loop variable. See https://github.com/bitcoin/bitcoin/pull/16702#discussion_r515608294.

Reply:
Yeah, some of them. I'll move them out.